### PR TITLE
Hide TabBar and NavigationBar when scrolling

### PR DESCRIPTION
**Pull request by noloman**
# Checklist
- [X] I have described what this PR contains

**Choose one of the following two options:**
- - [X] This PR does not introduce major changes
- - [ ] This PR introduces major changes, and I have consulted @buresdv, @jboi or @mormaer in the *Mlem Development Matrix room*

**Choose one of the following two options:**
- - [ ] This PR does not change the UI in any way
- - [X] This PR adds new UI elements / changes the UI, and I have attached pictures or videos of the new / changed UI elements

# Pull Request Information

## About this Pull Request
Addresses mlemgroup/mlem#38: **Scrolling up and than the scroll view shows and hides the TabBar and NavigationBar respectively**

## Screenshots and Videos
Simulator Screen Recording - iPhone 14 Pro - 2023-06-15 at 07 39 57

## Additional Context
Please let me know if there's a better way to do this! Thanks a lot!

### CODE REVIEW

**Comment by Que20** [comment]
Can't we keep a minimal navigation and bottom bar ? At least to cover up the safe areas and keep a good visibility over the status bar and the home indicator ?

**Comment by ShadowJonathan** [comment]
See the attached issue, we're trying to emulate Apollo's look

**Comment by Que20** [comment]
My bad. I did not use that feature.
That being said: one of the power features of Apollo was flexibility and how users can tweak the UI to fit their needs.

**Comment by noloman** [comment]
I'm a bit confused here sorry!

Should I then add an opaque background to the status bar or what should I
do?

**Comment by ShadowJonathan** [comment]
If that's possible, could you show a mock-up/screenshot here?

**Comment by noloman** [comment]
> If that's possible, could you show a mock-up/screenshot here?

Indeed, the status bar is transparent, but I'm not sure how to set a background color on it

Simulator Screenshot - iPhone 14 Pro - 2023-06-13 at 15 59 16

**Comment by WestonHanners** [comment]
Can we add this as an option? I have accessibility concerns about this being default behavior.

**Review by ShadowJonathan — CHANGES_REQUESTED** [review]
I also just noticed this is not a settings option, please make it one.

**Comment by Que20** [comment]
We have few options here: you can either completely hide the status bar (toggle the `.statusBar(hidden: true)` modifier) and get a real full screen experience, or place a view underneath it (using a simple `thinMaterial` view that ignores safe area with the modifier `.ignoresSafeArea()`) like Apollo does. 

Maybe the user could choose it's preferred full screen mode from settings ?
You can check on the PR https://github.com/buresdv/Mlem/pull/124 I made that introduces the setting options feature to allow users to get a list of selectable values in the settings.

**Comment by noloman** [comment]
> We have few options here: you can either completely hide the status bar (toggle the `.statusBar(hidden: true)` modifier) and get a real full screen experience, or place a view underneath it (using a simple `thinMaterial` view that ignores safe area with the modifier `.ignoresSafeArea()`) like Apollo does.
> 
> Maybe the user could choose it's preferred full screen mode from settings ? You can check on the PR #124 I made that introduces the setting options feature to allow users to get a list of selectable values in the settings.

Thanks! I already created the preferences, but I’ll adjust them to have these two options

**Comment by noloman** [comment]
hey, I'm not sure about what you exactly wanted it to be made with the `thinMaterial` but I came up with this:

Simulator Screen Recording - iPhone 14 Pro - 2023-06-15 at 07 39 57

**Comment by ShadowJonathan** [comment]
Looks good ^^

The only nit I have is that it doesn't stretch all the way to the bottom of the screen, but that's not blocking

**Comment by noloman** [comment]
Hey David,

I'm not very sure about what you exactly mean. When the feature is
disabled, I'm not changing anything.

**Comment by buresdv** [comment]
> I'm not very sure about what you exactly mean

I was wondering if when the hiding is enabled, the tab bar at the bottom of the screen and the toolbar at the top are transparent and blurry like when the hiding is disabled. I can't see that in the video.

> When the feature is disabled, I'm not changing anything.

Great 👌

**Comment by noloman** [comment]
Well I can try to upload a video somewhere else than GitHub (because of
size reasons) so you can check it out. How does it sound to you?

Thanks!

**Comment by noloman** [comment]
[Here](https://streamable.com/6j25ui) it is

File: Mlem/Views/Tabs/Settings/Components/Views/Appearance.swift [hunk 1]
```diff
@@ -149,7 +152,23 @@ struct AppearanceSettingsView: View {
                     isTicked: $shouldShowCommunityIcons
                 )
             }
-            
+            Section("Navigation")
+            {
+                SwitchableSettingsItem(
+                    settingPictureSystemName: "scroll.fill",
+                    settingPictureColor: .pink,
+                    settingName: "Hide TopBar and NavBar when scrolling",
```
Comment (mormaer) [inline Mlem/Views/Tabs/Settings/Components/Views/Appearance.swift:160]: Should we reword this to _"Hide top and bottom bars while scrolling"_, `TopBar` and `NavBar` sound like variable names 😉

**Review by mormaer — CHANGES_REQUESTED** [review]
Really like to see this in, but at the moment it has an issue where the user can get trapped in the settings tab, I can reproduce this by;
- Having the option disabled
- Scroll part of the way down a feed
- Switch to settings tab
- Enable this feature

At this point the tab bar is removed so I can't leave the settings screen 🙈 

https://github.com/mormaer/Mlem/assets/5231793/636dcd14-d887-4d1d-a09c-a2f9594ea9f0

is there a way to restrict the hiding of bars to only affect our feed tab?

File: Mlem/Views/Tabs/Settings/Components/Views/Appearance.swift (hunk 1 above)
Comment (noloman) [reply Mlem/Views/Tabs/Settings/Components/Views/Appearance.swift]: done, thanks!

**Comment by noloman** [comment]
> Really like to see this in, but at the moment it has an issue where the user can get trapped in the settings tab, I can reproduce this by;
> 
>     * Having the option disabled
> 
>     * Scroll part of the way down a feed
> 
>     * Switch to settings tab
> 
>     * Enable this feature
> 
> 
> At this point the tab bar is removed so I can't leave the settings screen 🙈
> bar_issue.mov
> 
> is there a way to restrict the hiding of bars to only affect our feed tab?

This is strange.. the hiding of the top and bottom bars in the settings tab didn't happen before, and now (perhaps after merging 135 commits from master and fixing conflicts?) is happening. I'll work it out, thanks!

**Comment by mormaer** [comment]
> This is strange.. the hiding of the top and bottom bars in the settings tab didn't happen before, and now (perhaps after merging 135 commits from master and fixing conflicts?) is happening. I'll work it out, thanks!

Yeah, I don't remember seeing it last time I ran this branch, hopefully it's a quick fix - thanks for looking into it.

If it's of any use it was ran on an iPhone 11 Pro / iOS 16.5

**Comment by noloman** [comment]
> > This is strange.. the hiding of the top and bottom bars in the settings tab didn't happen before, and now (perhaps after merging 135 commits from master and fixing conflicts?) is happening. I'll work it out, thanks!
> 
> Yeah, I don't remember seeing it last time I ran this branch, hopefully it's a quick fix - thanks for looking into it.
> 
> If it's of any use it was ran on an iPhone 11 Pro / iOS 16.5

Thanks, but this is very strange because I've been trying several times, and sometimes it hides the bars on the settings view, and sometimes it doesn't :(

**Comment by ShadowJonathan** [comment]
Heya @noloman, we have a new repo at https://github.com/mlemgroup/mlem, due to some shuffling around we couldn't get it migrated properly, but if you could reopen this PR there, and link it back here, then it can continue to be considered ^^
